jscomp/gentype_tests/typescript-react-example/src/JSXV4.gen.tsx

```diff
@@ -0,0 +1,48 @@
+/* TypeScript file generated from JSXV4.res by genType. */
+/* eslint-disable import/first */
+
+
+import * as React from 'react';
+
+// @ts-ignore: Implicit any on import
+import * as JSXV4BS__Es6Import from './JSXV4.bs';
+const JSXV4BS: any = JSXV4BS__Es6Import;
+
+// tslint:disable-next-line:interface-over-type-literal
+export type props<x,y> = {
+  readonly key?: string; 
+  readonly x: x; 
+  readonly y: y
+};
+
+// tslint:disable-next-line:interface-over-type-literal
+export type CompV4_props<x,y> = {
+  readonly key?: string; 
+  readonly x: x; 
+  readonly y: y
+};
+
+export const make: React.ComponentType<{
+  readonly key?: string; 
+  readonly x: string; 
+  readonly y: string
+}> = JSXV4BS.make;
+
+export const CompV4_make: React.ComponentType<{
+  readonly key?: string; 
+  readonly x: string; 
+  readonly y: string
+}> = JSXV4BS.CompV4.make;
+
+// tslint:disable-next-line:interface-over-type-literal
+export type Props = { readonly x: string; readonly y: string };
+
+export const CompV3_make: React.ComponentType<{ readonly x: string; readonly y: string }> = JSXV4BS.CompV3.make;
+
+export const CompV3: { make: React.ComponentType<{ readonly x: string; readonly y: string }> } = JSXV4BS.CompV3
+
+export const CompV4: { make: React.ComponentType<{
+  readonly key?: string; 
```

Comment:
@mattdamon108 why does the type of `props` include `key`?
Normally one never calls the function directly passing `key`, but uses instead the same mechanism as for refs.

Reply:
Unless the type `props` has the key, the type checking would not pass for the caller site.
```rescript
<Comp key="key" />
```

Reply:
JSX apply can treat key specially and simply add an explicit key:string type check.

Reply:
Let me take a look at the spec and make a concrete suggestion.

Reply:
Current:
```rescript
<Comp x key="7">
// is converted to
React.createElement(Comp.make, {x, key: "7"})
```

New:
```rescript
<Comp x key="7">
// is converted to
React.createElement(Comp.make, {x, key: ("7":string)})
```

Reply:
Does it make sense?

Reply:
@mattdamon108 I had a closer look and the core issue is the lack of an empty record expression. The `key` label is used as default label to indicate an empty record. At call time, I think we have no way to know what the props are. If we knew, one could take at least one optional prop and use that, but we don't know.

Is that your understanding too?

Reply:
FYI https://github.com/rescript-lang/rescript-compiler/issues/5623

Reply:
> Current:
> 
> ```rescript
> <Comp x key="7">
> // is converted to
> React.createElement(Comp.make, {x, key: "7"})
> ```
> 
> New:
> 
> ```rescript
> <Comp x key="7">
> // is converted to
> React.createElement(Comp.make, {x, key: ("7":string)})
> ```

Not sure I understand your intention.
Doesn't the `New` still needs the type props including `key`?

Reply:
> > Current:
> > ```rescript
> > <Comp x key="7">
> > // is converted to
> > React.createElement(Comp.make, {x, key: "7"})
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > New:
> > ```rescript
> > <Comp x key="7">
> > // is converted to
> > React.createElement(Comp.make, {x, key: ("7":string)})
> > ```
> 
> Not sure I understand your intention. Doesn't the `New` still needs the type props including `key`?

I think that's something one can work around.
But the lack of an empty record, one cannot work around I think https://github.com/rescript-lang/rescript-compiler/issues/5623

So for now I'll filter it out on the genType side, even if it's a bit of magic that ideally I would prefer not to put in there.

Reply:
Done here https://github.com/rescript-lang/rescript-compiler/pull/5619/commits/bfe431c5530b8e417021772175bc4eb5998c6460

Reply:
Side comment: I think you already know. The type props including `key` is needed, because the React.component has a type definition like this. https://github.com/rescript-lang/rescript-react/blob/master/src/React.res#L11:L19

Reply:
> Side comment: I think you already know. The type props including `key` is needed, because the React.component has a type definition like this. https://github.com/rescript-lang/rescript-react/blob/master/src/React.res#L11:L19

This is what I think one could work around, but the empty record, I don't think one can.